aiosmtpd/tests/test_starttls.py

```diff
@@ -1,236 +1,288 @@
-            code, response = client.docmd('AUTH PLAIN AHRlc3QAdGVzdA==')
-            assert_auth_invalid(self, code, response)
+# endregion
+
+
+# region #### Fixtures ###############################################################
+
+
+@pytest.fixture
+def tls_controller(get_handler) -> TLSController:
+    handler = get_handler()
+    controller = TLSController(handler)
+    controller.start()
+    Global.set_addr_from(controller)
+    #
+    yield controller
+    #
+    # Some test cases need to .stop() the controller inside themselves
+    # in such cases, we must suppress Controller's raise of AssertionError
+    # because Controller doesn't like .stop() to be invoked more than once
+    with suppress(AssertionError):
+        controller.stop()
+
+
+@pytest.fixture
+def tls_req_controller(get_handler) -> TLSController:
+    handler = get_handler()
+    controller = TLSRequiredController(handler)
+    controller.start()
+    Global.set_addr_from(controller)
+    #
+    yield controller
+    #
+    controller.stop()
+
+
+@pytest.fixture
+def auth_req_tls_controller(get_handler) -> RequireTLSAuthDecodingController:
+    handler = get_handler()
+    controller = RequireTLSAuthDecodingController(handler)
+    controller.start()
+    Global.set_addr_from(controller)
+    #
+    yield controller
+    #
+    controller.stop()
+
+
+@pytest.fixture
+def simple_controller() -> SimpleController:
+    handler = Sink()
+    controller = SimpleController(handler)
+    controller.start()
+    Global.set_addr_from(controller)
+    #
+    yield controller
+    #
+    controller.stop()
+
+
+@pytest.fixture
+def client() -> SMTP:
+    with SMTP(*Global.SrvAddr) as client:
+        yield client
+
+
+# endregion
+
+
+def test_disabled_tls(simple_controller, client):
+    code, _ = client.ehlo("example.com")
+    assert code == 250
+    resp = client.docmd("STARTTLS")
+    assert resp == (454, b"TLS not available")
+
+
+@pytest.mark.usefixtures("tls_controller")
+class TestStartTLS:
+    @pytest.mark.handler_data(class_=ReceivingHandler)
+    def test_starttls(self, tls_controller, client):
+        sender = "sender@example.com"
+        recipients = ["rcpt1@example.com"]
+        code, _ = client.ehlo("example.com")
+        assert code == 250
+        assert "starttls" in client.esmtp_features
+        resp = client.starttls()
+        assert resp == S.S220_READY_TLS
+        client.send_message(MIMEText("hi"), sender, recipients)
+        handler: ReceivingHandler = tls_controller.handler
+        assert len(handler.box) == 1
+        assert handler.box[0].mail_from == sender
+        assert handler.box[0].rcpt_tos == recipients
+
+    def test_starttls_quit(self, tls_controller, client):
+        code, _ = client.ehlo("example.com")
+        assert code == 250
+        resp = client.starttls()
+        assert resp == S.S220_READY_TLS
+        resp = client.quit()
+        assert resp == S.S221_BYE
+        client.close()
```

Comment:
This is a new test because sometimes PyPy3 failed coverage. See comment of commit 1847819d